libpromises/eval_context.c

```diff
@@ -1443,6 +1450,8 @@ static bool EvalContextClassPut(EvalContext *ctx, const char *ns, const char *na
         return false;
     }
 
+    StringSetAdd(ctx->all_classes, SafeStringDuplicate(name));
```

Comment:
Another solution would be to start with a non-NULL `StringSet` inside `EvalContext` and destroy it and make it NULL when you are sure that the classes logging option is disabled.

The line in this file would just change to:
```c
if (ctx->all_classes != NULL)
     StringSetAdd(ctx->all_classes, SafeStringDuplicate(name)); 
```